experiments/storm/src/main/scala/io/gearpump/experiments/storm/producer/StormSpoutOutputCollector.scala

```diff
@@ -36,7 +45,43 @@ private[storm] class StormSpoutOutputCollector(collector: StormOutputCollector)
     throw throwable
   }
 
-  override def emitDirect(taskId: Int, streamId: String, values: JList[AnyRef], messageId: scala.Any): Unit = {
+  override def emitDirect(taskId: Int, streamId: String, values: JList[AnyRef], messageId: Object): Unit = {
+    setPendingOrAck(messageId)
     collector.emitDirect(taskId, streamId, values)
   }
+
+
+  def ackPendingMessage(checkpointClock: TimeStamp): Unit = {
+    pendingMessage.foreach { case PendingMessage(id, messageTime, _) =>
+      if (messageTime <= checkpointClock) {
+        spout.ack(id)
```

Comment:
Are you sure calling spout.ack in same thread of emit is consistent with native storm?